File: skyspark-scala/src/main/scala/com/github/dkanellis/skyspark/scala/api/algorithms/bnl/Merger.scala
package com.github.dkanellis.skyspark.scala.api.algorithms.bnl

import com.github.dkanellis.skyspark.scala.api.algorithms.Point
import org.apache.spark.rdd.RDD

/**
  * This is responsible for the Merging part of the MR-BNL algorithm.
  *
  * @param bnlAlgorithm The core BNL algorithm
  */
private[bnl] class Merger(private val bnlAlgorithm: BnlAlgorithm) extends Serializable {

  private[bnl] def merge(flagsWithPoints: RDD[(Flag, Point)]): RDD[Point] = {
    val inSinglePartition = flagsWithPoints
      .coalesce(1)
      .glom()

    inSinglePartition.flatMap(bnlAlgorithm.computeSkylinesWithPreComparison(_))
  }
}

File: skyspark-java/src/main/java/com/akikanellis/skyspark/api/algorithms/bnl/Merger.java
package com.akikanellis.skyspark.api.algorithms.bnl;

import com.akikanellis.skyspark.api.algorithms.Point;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import scala.Tuple2;

import java.io.Serializable;
import java.util.List;

class Merger implements Serializable {
    private final BnlAlgorithm bnlAlgorithm;

    Merger(BnlAlgorithm bnlAlgorithm) { this.bnlAlgorithm = bnlAlgorithm; }

    JavaRDD<Point> merge(JavaPairRDD<Flag, Point> flagPoints) {
        JavaRDD<List<Tuple2<Flag, Point>>> inSinglePartition = flagPoints
                .coalesce(1)
                .glom();

        return inSinglePartition.flatMap(bnlAlgorithm::computeSkylinesWithPreComparison);
    }
}
